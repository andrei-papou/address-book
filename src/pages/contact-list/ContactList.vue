<template>

	<div>
		<div :class="{blurred: backgroundBlur}">
			<top-bar></top-bar>
			<div class="container">
				<b-row class="heading-row">
					<b-col>
						<h4>Clients</h4>
					</b-col>
					<b-col class="text-right">
						<router-link :to="{name: 'create-contact'}">
							<b-button><i class="fa fa-plus" aria-hidden="true"></i>&nbsp; Add client</b-button>
						</router-link>
					</b-col>
				</b-row>
				<b-table outlined
						 :items="allContacts"
						 :fields="fields"
						 :sort-by.sync="sortBy" 
						 :sort-desc.sync="sortDesc">
					<template slot="edit" slot-scope="data">
						<router-link :to="{name: 'edit-contact', params: {id: data.item.name}}">
							<i class="fa fa-pencil edit-link" aria-hidden="true"></i>
						</router-link>
					</template>
				</b-table>
			</div>
		</div>

		<router-view :set-background-blur="setBackgroundBlur"></router-view>

	</div>

</template>

<script>

import { mapGetters, mapActions } from 'vuex';
import TopBar from '@/components/TopBar';

export default {
  	components: {
  		TopBar,
  	},
  	computed: {
    	...mapGetters(['allContacts']),
  	},
  	methods: {
  		...mapActions(['getContacts']),

  		setBackgroundBlur(val) {
  			this.backgroundBlur = val;
  		}
  	},

  	data() {
  		return {
  			sortBy: null,
  			sortDesc: false,
  			fields: [
				{ key: 'name', sortable: true },
				{ key: 'phone', sortable: true },
				{ key: 'birthday', sortable: true },
				{ key: 'edit', label: ' ', tdClass: 'text-right' },
  			],
  			backgroundBlur: false,
  		};
  	},
  
  	mounted() {
    	this.getContacts();
  	},
}

</script>

<style>

</style>
