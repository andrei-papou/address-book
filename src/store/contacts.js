import moment from 'moment';
import Vue from 'vue';
import contactsApi from '@/api/contacts';
import {
	CONTACTS_LIST,
	CONTACTS_EDIT,
	CONTACTS_CREATE,
} from './mutation-types';


const state = {
	all: [],
};

export const actions = {
	getContacts({ commit }) {
		return contactsApi.getContactList()
			.then(response => commit(CONTACTS_LIST, response))
			.catch(error => console.error(error));
	},

	createContact({ commit }, contactData) {
		return contactsApi.createContact(contactData)
			.then(() => commit(CONTACTS_CREATE, contactData))
			.catch(err => console.log(err));
	},

	updateContact({ commit }, updatedObj) {
		commit(CONTACTS_EDIT, updatedObj);
	},
};

export const mutations = {
	[CONTACTS_LIST](state, contacts) {
		contacts.forEach(c => c.birthday = moment(c.birthday, 'YYYY-MM-DD').format('MM/DD/YYYY'));
		state.all = contacts;
	},

	[CONTACTS_CREATE](state, newObj) {
		state.all.push(newObj);
	},

	[CONTACTS_EDIT](state, updatedObj) {
		const index = state.all.findIndex(c => c.name === updatedObj.name);
		Vue.set(state.all, index, Object.assign({}, state.all[index], updatedObj));
	}
};

const getters = {
	allContacts: state => state.all,
	getContactByName: state => name => state.all.find(c => c.name === name),
};

export default {
	state,
	actions,
	mutations,
	getters
};
