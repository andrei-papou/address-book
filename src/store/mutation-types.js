export const CONTACTS_LIST = 'contacts-list';
export const CONTACTS_CREATE = 'contacts-create';
export const CONTACTS_EDIT = 'contacts-edit';
