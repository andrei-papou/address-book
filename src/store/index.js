import Vuex from 'vuex';
import Vue from 'vue';
import contacts from './contacts';


Vue.use(Vuex);

export default new Vuex.Store({
	modules: {
		contacts
	}
});
