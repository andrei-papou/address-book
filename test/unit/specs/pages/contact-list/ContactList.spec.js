import Vue from 'vue';
import store from '@/store';
import router from '@/routes';
import ContactList from '@/pages/contact-list/ContactList';
import { mockFetch } from '../../../mock.js';

describe('ContactList.vue', () => {

	let vm;
	const Constructor = Vue.extend(ContactList);

	beforeEach(() => {
		mockFetch({ ok: true, json: () => Promise.resolve([]) });
	});

	it('should render correct dom elements', () => {
		vm = new Constructor({ store, router }).$mount();
		expect(vm.$el.querySelector('.container')).toBeTruthy();
		expect(vm.$el.querySelector('a[href="/create"]')).toBeTruthy();
		expect(vm.$el.querySelector('tr')).toBeTruthy();
		expect(vm.$el).toMatchSnapshot();
	});

	it('should contain setBackgroundBlur method with correct behavior', () => {
		vm = new Constructor({ store, router }).$mount();
		expect(typeof vm.setBackgroundBlur).toEqual('function');
		expect(vm.backgroundBlur).toBeFalsy();
		vm.setBackgroundBlur(true);
		expect(vm.backgroundBlur).toBeTruthy();
	});

	it('should contain mounted hook with correct behavior', () => {
		vm = new Constructor({ store, router })
		vm.getContacts = jest.fn();
		vm.$mount();
		expect(vm.getContacts).toHaveBeenCalled();
	});

});
