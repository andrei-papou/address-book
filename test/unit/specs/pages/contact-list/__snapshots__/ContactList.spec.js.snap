// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList.vue should render correct dom elements 1`] = `
<div>
  <div
    class=""
  >
    <header
      class="app-header"
    >
      <div
        class="container"
      >
        <img
          src="@/assets/images/logo.png"
        />
      </div>
    </header>
     
    <div
      class="container"
    >
      <div
        class="row heading-row"
      >
        <div
          class="col"
        >
          <h4>
            Clients
          </h4>
        </div>
         
        <div
          class="text-right col"
        >
          <a
            class=""
            href="/create"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              <i
                aria-hidden="true"
                class="fa fa-plus"
              />
              Â  Add client
            </button>
          </a>
        </div>
      </div>
       
      <table
        aria-busy="false"
        aria-colcount="4"
        class="table b-table border"
      >
        <!---->
        <!---->
        <thead
          class=""
        >
          <tr>
            <th
              aria-colindex="1"
              aria-label="Click to sort Descending"
              class="sorting"
              tabindex="0"
            >
              Name
            </th>
            <th
              aria-colindex="2"
              aria-label="Click to sort Descending"
              class="sorting"
              tabindex="0"
            >
              Phone
            </th>
            <th
              aria-colindex="3"
              aria-label="Click to sort Descending"
              class="sorting"
              tabindex="0"
            >
              Birthday
            </th>
            <th
              aria-colindex="4"
              class=""
            >
               
            </th>
          </tr>
        </thead>
        <!---->
        <tbody
          class=""
        >
          <!---->
          <!---->
          <!---->
        </tbody>
      </table>
    </div>
  </div>
   
  <!---->
</div>
`;
